### PR TITLE
OpenSea API fix + Custom API support + Auto Retry + Retry Button

### PR DESCRIPTION
**Pull request by trkaplan**
OpenSea changed the API, instead of `offset` it now uses `cursor` and ask for an API KEY to fetch the data. 
I've made some changes to fix this issue plus add a auto-retry feature in case of the request failure. Plus an Retry button if it still fails after n times of retry it shows a Retry Button
ss-opensea-api-fix-new-props
ss-opensea-api-fix-retry-button

### CODE REVIEW

**Comment by bkrem** [comment]
> OpenSea changed the API, instead of offset it now uses cursor and ask for an API KEY to fetch the data.

Oh man not again 🤦‍♂️ the amount of breaking changes these guys ship on their `/v1/` API path is ridiculous.

Thanks for the work you did here, looking really good from my first look through the changes 💯  

You're my hero for setting up an API proxy for the Opensea API 🙏  I knew that what was really needed here (not just a prop for the API key itself), but they never gave me an API key I could work with.

Regarding the retry logic, is this necessary in your experience or handling a rare occurrence, i.e. are you seeing the need for retries? In my experience working with the Opensea API this hasn't been an issue, but curious to hear why there's explicit retry logic here.

**Comment by trkaplan** [comment]
> 

You are welcome, I forgot to thank you first sorry :) 
Here is my answers:
I added the auto-retry mechanism because I was using a shared api key since I couldn't get one too. And I think some bots were using the same API key and I was receiving 429 (`Request was throttled. Expected available in 1 second.`) most of the time. I could implement this on the server side and cache the latest succesfull response for each request but I didn't have time.

So the quickest solution was adding this mechanism to the client side. But then I found another API key (is not used by bots I think) and It works well (haven't seen a 429 error yet)

API proxy service works well not but I'm going to limit it for only one owner address (since this API is used in production and it can be abused).

So we can omit the CityDAO story and remove this API address. I wanted to share this with you ASAP to get your feedback so I didn't think more about this proxy api situation.

I'd be happy if you have tips on the caching mechanism for the [opensea-proxy-api](https://github.com/trkaplan/opensea-api-proxy) which is ready-to-run on vercel. (of course if you have an API key.)

If you can add your comments to the code, I can update the PR.

File: README.md [hunk 1]
```diff
@@ -90,6 +108,21 @@ interface NftGalleryProps {
...
+   */
+  isProxyApi?: boolean;
+
+  /**
+   * Set custom API URL.
+   */
+  apiUrl?: string;
+
+  /**
+   * Auto retry when API request failed.
```
Comment (bkrem) [inline README.md]:
```suggestion
   * Auto retry (10 times by default) after an API request failed.
```

File: src/api.ts [hunk 2]
```diff
@@ -1,7 +1,12 @@
-import { OpenseaAsset } from './types/OpenseaAsset';
+import { OpenseaAssetsAndNextCursor } from './types/OpenseaAsset';
 
 export const OPENSEA_API_OFFSET = 50;
+const OPENSEA_URL = 'https://api.opensea.io';
+const ENS_GRAPH_URL = 'https://api.thegraph.com/subgraphs/name/ensdomains/ens';
+const MAX_AUTO_RETRY_ATTEMPT = 20;
```
Comment (bkrem) [inline src/api.ts]: Setting auto retry to the max indicated in the prop comment.

```suggestion
const MAX_AUTO_RETRY_ATTEMPT = 10;
```

File: src/NftGallery.tsx [hunk 3]
```diff
@@ -225,11 +278,31 @@ export const NftGallery: React.FC<NftGalleryProps> = ({
...
+        currentCursor
+      );
     }
-  }, [showcaseMode, ownerAddress, openseaApiKey, currentOffset]);
+  }, [
+    showcaseMode,
+    ownerAddress,
+    openseaApiKey,
+    isProxyApi,
+    apiUrl,
```
Comment (bkrem) [inline src/NftGallery.tsx:301]: Missing `autoRetry` as a hook dependency:

```suggestion
    apiUrl,
    autoRetry,
```

**Comment by bkrem** [comment]
I noticed that there's currently an infinite loop when trying to use the new props with `showcaseMode`, e.g. when setting the following in the storybook and trying to view the "Fidenza Showcase" story:

```ts
ThreeArrowsCapitalShowAll.args = {
  ownerAddress: '0x2e675eeae4747c248bfddbafaa3a8a2fdddaa44b',
  apiUrl: 'http://opensea-api-proxy.vercel.app',
  isProxyApi: true,
  darkMode: true,
  autoRetry: true,
} as NftGalleryProps;
```

I did some debugging and this was due to the `cursor` in the local scope not being updated the way `currentOffset` was previously, which is needed to paginate automatically and then break the loop once all assets are fetched. 
Here's a commit with a fix: https://github.com/bkrem/react-nft-gallery/commit/107378a6319b4e85b8d0a0637e238c437352f008

**Comment by bkrem** [comment]
Thanks for the tweaks here @trkaplan 🙏

Haven’t forgotten about this, simply a busy week with other responsibilities for me.

I’m planning to merge this into a branch by end of tomorrow, make a few more small tweaks and ship a new version if everything is working right 👍

**Comment by trkaplan** [comment]
> I noticed that there's currently an infinite loop when trying to use the new props with `showcaseMode`, e.g. when setting the following in the storybook and trying to view the "Fidenza Showcase" story:
> 
> ```ts
> ThreeArrowsCapitalShowAll.args = {
>   ownerAddress: '0x2e675eeae4747c248bfddbafaa3a8a2fdddaa44b',
>   apiUrl: 'http://opensea-api-proxy.vercel.app',
>   isProxyApi: true,
>   darkMode: true,
>   autoRetry: true,
> } as NftGalleryProps;
> ```
> 
> I did some debugging and this was due to the `cursor` in the local scope not being updated the way `currentOffset` was previously, which is needed to paginate automatically and then break the loop once all assets are fetched. Here's a commit with a fix: [107378a](https://github.com/bkrem/react-nft-gallery/commit/107378a6319b4e85b8d0a0637e238c437352f008)

Awesome waiting for the next version:) 
These new parameters should be added to Storybook as well, otherwise Storybook shows an empty screen. 
Ping me If you you don't have time and if you are able to assign a task to me.

**Pull request merged**